File: app/src/main/java/com/jbrunton/mymovies/api/repositories/HttpGenresRepository.java
package com.jbrunton.mymovies.api.repositories;

import com.jbrunton.entities.Genre;
import com.jbrunton.entities.GenresRepository;
import com.jbrunton.mymovies.api.resources.GenresResponse;
import com.jbrunton.mymovies.api.services.MovieService;

import java.util.List;

import io.reactivex.Observable;

public class HttpGenresRepository implements GenresRepository {
    private final MovieService service;

    public HttpGenresRepository(MovieService service) {
        this.service = service;
    }

    @Override public Observable<List<Genre>> genres() {
        return service.genres()
                .map(GenresResponse::toCollection);
    }
}

File: networking/src/main/java/com/jbrunton/mymovies/networking/repositories/HttpGenresRepository.kt
package com.jbrunton.mymovies.networking.repositories

import com.jbrunton.mymovies.entities.models.Genre
import com.jbrunton.mymovies.entities.repositories.*
import com.jbrunton.mymovies.networking.services.MovieService


class HttpGenresRepository(private val service: MovieService) : GenresRepository {
    override suspend fun genres(): DataStream<List<Genre>> {
        return buildDataStream {
            service.genres().toCollection()
        }
    }
}
